### PR TITLE
Adds a how section to the readme

### PR DESCRIPTION
**Pull request by RobStallion**
#3 
adds sections on operations and array to readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,9 +45,228 @@ for example:
 4 - Put other slice on top of filling 🥪
```
Comment (Cleop) [inline README.md:45]: What is the  🥪 meant to be? Is it meant to be an emoji? 
image

File: README.md [hunk 2]
```diff
@@ -45,9 +45,228 @@ for example:
...
 ```
 
-
 In computing an algorithm is the same thing. Just a set of steps for a computer
 program to accomplish a task.
 
 There are often a number of ways your could accomplish a task so learning to
 pick the right algorithm is a key.
```
Comment (Cleop) [inline README.md]:
```suggestion
 pick the right algorithm is key.
```

File: README.md [hunk 3]
```diff
@@ -45,9 +45,228 @@ for example:
...
+## *How*?
+
+Most examples of code in this README will be in elixir. If you would like to
+learn a little more about elixir, head over to
+[learn-elixir](https://github.com/dwyl/learn-elixir).
+
+## Data Structures
+
+As we mentioned earlier, data structures refer to how data is organised. This
+organisation is not just for organisations sake however. The data structures you
```
Comment (Cleop) [inline README.md]:
```suggestion
organisation is not just for organisation's sake however. The data structures you
```

File: README.md [hunk 4]
```diff
@@ -45,9 +45,228 @@ for example:
...
+
+An [array](https://en.wikipedia.org/wiki/Array_data_structure) is just a list of
+data elements.
+
+```
+array = ["one", "two", "three", "four", "five"]
+```
+
+#### Reading
+> The following examples are not 'exactly' how the computer store arrays in
```
Comment (Cleop) [inline README.md]:
```suggestion
> The following examples are not 'exactly' how the computer stores arrays in
```

File: README.md [hunk 5]
```diff
@@ -45,9 +45,228 @@ for example:
...
+memory. The computer actually stores all the elements of an array together in
+memory, with each cell of the computers memory holding one elements value. The
```
Comment (Cleop) [inline README.md]:
```suggestion
memory, with each cell of the computer's memory holding one element's value. The
```

File: README.md [hunk 6]
```diff
@@ -45,9 +45,228 @@ for example:
...
+computer records the memory address of where that array started and uses it to
+jump straight to that spot in memory. How the computer store the information is
```
Comment (Cleop) [inline README.md]:
```suggestion
jump straight to that spot in memory. How the computer stores the information is
```

File: README.md [hunk 7]
```diff
@@ -45,9 +45,228 @@ for example:
...
+not the key point here so we will not go into much detail here.
+
+Reading from an array takes one step. Each cell in an array is given an index.
```
Comment (Cleop) [inline README.md]:
```suggestion
**Reading from an array takes one step**. Each cell in an array is given an index.
```

I think this should be bold as it's the main take away from this bit.

File: README.md [hunk 8]
```diff
@@ -45,9 +45,228 @@ for example:
...
+
+So for using the array above, say we want to get the element from the 2nd index.
+The computer knows that the first index is index 0 and that index 2 is exactly 2
+over from index 0. With this the computer can jump right to the 2nd and then
+grab the data from that point.
+
+#### Searching
+If we wanted to check if a value exists in this array then we would use the
+search operation. Say we wanted to check if the value `"six"` was in the array.
+We can just glance at the array and clearly see that it is not but the computer
```
Comment (Cleop) [inline README.md]:
```suggestion
We can just glance at the array and clearly see that it is not, but the computer
```

File: README.md [hunk 9]
```diff
@@ -45,9 +45,228 @@ for example:
...
+does not have this skill. The computer needs to access every element
+individually and check if the value it get's back is `"six"`. If it finds a
+match it will stop and return true else it will continue till there are no more
```
Comment (Cleop) [inline README.md]:
```suggestion
match it will stop and return `true` otherwise it will continue until there are no more
```

File: README.md [hunk 10]
```diff
@@ -45,9 +45,228 @@ for example:
...
+elements, at which point it will return false.
```
Comment (Cleop) [inline README.md]:
```suggestion
elements, at which point it will return `false`.
```

File: README.md [hunk 11]
```diff
@@ -45,9 +45,228 @@ for example:
...
+
+Let's count the number of steps that searching our array for the value "six"
```
Comment (Cleop) [inline README.md]:
```suggestion
Let's count the number of steps that searching our array for the value `"six"`
```

File: README.md [hunk 12]
```diff
@@ -45,9 +45,228 @@ for example:
...
+would take. The computer starts by checking index 0 and sees that the value it
+contains is `"one"` so moves on to the next index. At index 1 it checks the
+value and sees that it is `"two"`. This is not what we are looking for so the
+computer again moves on....
+(the computer repeats these steps until it gets to the end of the array).
+
+In total the computer had to check 5 elements before it could be sure that the
```
Comment (Cleop) [inline README.md]:
```suggestion
In total **the computer had to check 5 elements before it could be sure that the
```

File: README.md [hunk 13]
```diff
@@ -45,9 +45,228 @@ for example:
...
+value was not there so this operation took 5 steps to complete. If the array
```
Comment (Cleop) [inline README.md]:
```suggestion
value was not there so this operation took 5 steps to complete**. If the array
```

File: README.md [hunk 14]
```diff
@@ -45,9 +45,228 @@ for example:
...
+had a million values then the operation could potentially take million steps to
```
Comment (Cleop) [inline README.md]:
```suggestion
had a million values then the operation **could potentially take a million steps** to
```

File: README.md [hunk 15]
```diff
@@ -45,9 +45,228 @@ for example:
...
+complete. However, if the element we were searching for was the first element in
+the array, then the operation would only take 1 step. The number of steps is
+dependent on where or if the element is in the array.
```
Comment (Cleop) [inline README.md]:
```suggestion
dependent on where or if the element is in the array.**
```

File: README.md [hunk 16]
```diff
@@ -45,9 +45,228 @@ for example:
...
+contains is `"one"` so moves on to the next index. At index 1 it checks the
+value and sees that it is `"two"`. This is not what we are looking for so the
+computer again moves on....
+(the computer repeats these steps until it gets to the end of the array).
+
+In total the computer had to check 5 elements before it could be sure that the
+value was not there so this operation took 5 steps to complete. If the array
+had a million values then the operation could potentially take million steps to
+complete. However, if the element we were searching for was the first element in
+the array, then the operation would only take 1 step. The number of steps is
```
Comment (Cleop) [inline README.md]:
```suggestion
the array, then the operation would only take 1 step. **The number of steps is
```

File: README.md [hunk 17]
```diff
@@ -45,9 +45,228 @@ for example:
...
+dependent on where or if the element is in the array.
+
+This type of search is known as a linear search. There are other, more complex
+types of search as well but this is just a basic search that will work will all
```
Comment (Cleop) [inline README.md]:
```suggestion
types of search as well but this is just a basic search that will work with all
```

File: README.md [hunk 18]
```diff
@@ -45,9 +45,228 @@ for example:
...
+(the computer repeats these steps until it gets to the end of the array).
+
+In total the computer had to check 5 elements before it could be sure that the
+value was not there so this operation took 5 steps to complete. If the array
+had a million values then the operation could potentially take million steps to
+complete. However, if the element we were searching for was the first element in
+the array, then the operation would only take 1 step. The number of steps is
+dependent on where or if the element is in the array.
+
+This type of search is known as a linear search. There are other, more complex
```
Comment (Cleop) [inline README.md]:
```suggestion
This type of search is known as a **linear search**. There are other, more complex
```

File: README.md [hunk 19]
```diff
@@ -45,9 +45,228 @@ for example:
...
+array types.
+
+#### Inserting and deleting
+Inserting and deleting from an array work fairly similarly to one another so
+I'll cover both here.
+
+Say we want to insert a value into our array. Similar to how the number of
+steps in searching is dependent on where the value is in the array, inserting
+depends on where you want to insert the element into the array. If you want to
+insert an element on the the end of an array then this is considered the 'best
```
Comment (Cleop) [inline README.md]:
```suggestion
insert an element onto the end of an array then this is considered the 'best
```

File: README.md [hunk 20]
```diff
@@ -45,9 +45,228 @@ for example:
...
+case scenario' as it only take the computer one step.
+
+The number of steps increase when you want to insert an element anywhere else in
+the array. Say for example you want to insert the `"ten"` into our array from
```
Comment (Cleop) [inline README.md]:
```suggestion
the array. Say for example you want to insert the value`"ten"` into our array near the 
```

File: README.md [hunk 21]
```diff
@@ -45,9 +45,228 @@ for example:
...
+earlier at index 2. Currently, index 2 has a value. It has a the value
```
Comment (Cleop) [inline README.md]:
```suggestion
beginning at index 2. But index 2 already has a value, it has the value
```

File: README.md [hunk 22]
```diff
@@ -45,9 +45,228 @@ for example:
...
+step 3 - shifting value "four" from index 4 to index 3
+step 4 - shifting value "five" from index 5 to index 4
+```
+
+The worst case for deleting, just like inserting, is to delete the first
+element from the array.
+
+### Set (array-based set)
+
+An array based set is very similar to an array but with one key difference, it
```
Comment (Cleop) [inline README.md]:
```suggestion
An array based set is very similar to an array but with one key difference, **it
```

File: README.md [hunk 23]
```diff
@@ -45,9 +45,228 @@ for example:
...
+never allows duplicate values to be inserted.
```
Comment (Cleop) [inline README.md]:
```suggestion
never allows duplicate values to be inserted.**
```

File: README.md [hunk 24]
```diff
@@ -45,9 +45,228 @@ for example:
...
+
+Sets come in handy when you need to make sure that you have no duplicate
+data.
+
+The operations that we used on our array all work in the same way on our set,
+with the exception of insert.
+
+When we call insert on our array based set it does work similarly to our example
+above but, before it can insert, it first needs to check every cell to make sure
+the value we want to insert if not already in the set.
```
Comment (Cleop) [inline README.md]:
```suggestion
the value we want to insert is not already in the set.
```

**Review by Cleop — COMMENTED** [review]
Really nicely written @RobStallion, just a few typos and small enhancements using bold. A really educational read 🤓

File: README.md [hunk 25]
```diff
@@ -49,7 +49,7 @@ In computing an algorithm is the same thing. Just a set of steps for a computer
 program to accomplish a task.
 
 There are often a number of ways your could accomplish a task so learning to
-pick the right algorithm is a key.
+ pick the right algorithm is key.
```
Comment (nelsonic) [inline README.md]: "select an _effective_ algorithm is key."

"the right" implies that there is only _one_ "right answer".
when it comes to algorithms, unless we apply a mathematical proof, 
we don't know if something is "ideal", so best describe it as "effective" for current purposes.

> see: "How do we right and wrong?": https://youtu.be/kIIKOXFTeBA

File: README.md (hunk 1 above)
Comment (RobStallion) [reply README.md:45]: It is meant to be a sandwich emoji 😂. I guess it is not clear then

File: README.md (hunk 25 above)
Comment (RobStallion) [reply README.md]: Good point. I'll update that in the next commit 👍 
Thanks for the spot 🔍

**Comment by RobStallion** [comment]
@Cleop Thanks for going through this. You really went through it with a fine tooth comb 👍 🥇

File: README.md (hunk 1 above)
Comment (Cleop) [reply README.md:45]: I think it's because I haven't updated my OS, my bad

**Comment by nelsonic** [comment]
@RobStallion would be _amaze_ to get this PR finished this week so others can start contributing. 😉 
(_hope you are feeling **much** better!_) ❤️

**Comment by RobStallion** [comment]
@nelsonic Will do. Sorry for the delay 😖

**Comment by nelsonic** [comment]
@RobStallion Don't worry about delays, you had/have plenty of other work to do. ✅ 
I'm _stoked_ that you are taking the time to summarise your learning/knowledge on data structures. 🎉

**Comment by RobStallion** [comment]
@nelsonic I do not think that this readme is done (not by a long shot) but I think that this is a good starting point. I was thinking to add something along those lines to the readme itself but didn't know if that was the best idea. Any thoughts/feedback on it would be super cool 👍

**Review by nelsonic — APPROVED** [review]
@RobStallion thanks for adding the "How?" section! 🎉 
I _agree_ that is still a _lot_ we can add, but this is _miles_ better than a _blank_ section. 😉

**Comment by RobStallion** [comment]
@nelsonic fair point 😂 

I'll continue learning more about this and try to improve it as I go 👍

**Pull request merged**